doc/man/man3/parser.3
.\" Copyright (c) 1993 by Thomas Koenig (ig25@rz.uni-karlsruhe.de)
.\"
.\" %%%LICENSE_START(VERBATIM)
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Modified Sat Jul 24 19:00:59 1993 by Rik Faith (faith@cs.unc.edu)
.\" Clarification concerning realloc, iwj10@cus.cam.ac.uk (Ian Jackson), 950701
.\" Documented MALLOC_CHECK_, Wolfram Gloger (wmglo@dent.med.uni-muenchen.de)
.\" 2007-09-15 mtk: added notes on malloc()'s use of sbrk() and mmap().
.\"
.\" FIXME . Review http://austingroupbugs.net/view.php?id=374
.\" to see what changes are required on this page.
.\"
.TH PARSER 3  2017-09-15 "GNU" "Blastpit Manual"
.SH NAME
parser \- parse a blastpit command packet
.SH SYNOPSIS
.nf
.B #include <new.h>
.B #include <parser.h>
.PP
.BI "t_parser *new(parser, char " "*pkt" );
.BI "int prsr_getCmd(t_parser " "*prsr" );
.BI "int prsr_getInt(t_parser " "*prsr" ", int " "index" );
.BI "long prsr_getLong(t_parser " "*prsr" ", int " "index" );
.BI "float prsr_getFloat(t_parser " "*prsr" ", int " "index" );
.BI "double prsr_getDouble(t_parser " "*prsr" ", int " "index" );
.fi
.PP
.SH DESCRIPTION
.PP
The
.BR parser
class extracts
.B Blastpit
commands and data from the
.I pkt
buffer.
.PP
The
.BR prse_getCmd ()
method returns a value of type
.IR BpCommand , 
or -1 if the buffer could not be parsed.
.PP
The methods
.BR prsr_getChar (),
.BR prsr_getFloat (),
and
.BR prsr_getInt ()
return the respective datatypes at position
.IR index ,
or -1 if the datatype cannot be parsed.
.PP
The
.BR reallocarray ()
function changes the size of the memory block pointed to by
.I ptr
to be large enough for an array of
.I nmemb
elements, each of which is
.I size
bytes.
It is equivalent to the call
.PP
.in +4n
    realloc(ptr, nmemb * size);
.in
.PP
However, unlike that
.BR realloc ()
call,
.BR reallocarray ()
fails safely in the case where the multiplication would overflow.
If such an overflow occurs,
.BR reallocarray ()
returns NULL, sets
.I errno
to
.BR ENOMEM ,
and leaves the original block of memory unchanged.
.SH RETURN VALUE
The
.BR malloc ()
and
.BR calloc ()
functions return a pointer to the allocated memory,
which is suitably aligned for any built-in type.
On error, these functions return NULL.
NULL may also be returned by a successful call to
.BR malloc ()
with a
.I size
of zero,
or by a successful call to
.BR calloc ()
with
.I nmemb
or
.I size
equal to zero.
.PP
The
.BR free ()
function returns no value.
.PP
The
.BR realloc ()
function returns a pointer to the newly allocated memory, which is suitably
aligned for any built-in type and may be different from
.IR ptr ,
or NULL if the request fails.
If
.I size
was equal to 0, either NULL or a pointer suitable to be passed to
.BR free ()
is returned.
If
.BR realloc ()
fails, the original block is left untouched; it is not freed or moved.
.PP
On success, the
.BR reallocarray ()
function returns a pointer to the newly allocated memory.
On failure,
it returns NULL and the original block of memory is left untouched.
.SH ERRORS
.BR calloc (),
.BR malloc (),
.BR realloc (),
and
.BR reallocarray ()
can fail with the following error:
.TP
.B ENOMEM
Out of memory.
Possibly, the application hit the
.BR RLIMIT_AS
or
.BR RLIMIT_DATA
limit described in
.BR getrlimit (2).
.SH NOTES
By default, Linux follows an optimistic memory allocation strategy.
This means that when
.BR malloc ()
returns non-NULL there is no guarantee that the memory really
is available.
In case it turns out that the system is out of memory,
one or more processes will be killed by the OOM killer.
For more information, see the description of
.IR /proc/sys/vm/overcommit_memory
and
.IR /proc/sys/vm/oom_adj
in
.BR proc (5),
and the Linux kernel source file
.IR Documentation/vm/overcommit-accounting .
.PP
Normally,
.BR malloc ()
allocates memory from the heap, and adjusts the size of the heap
as required, using
.BR sbrk (2).
When allocating blocks of memory larger than
.B MMAP_THRESHOLD
bytes, the glibc
.BR malloc ()
implementation allocates the memory as a private anonymous mapping using
.BR mmap (2).
.B MMAP_THRESHOLD
is 128\ kB by default, but is adjustable using
.BR mallopt (3).
Prior to Linux 4.7
allocations performed using
.BR mmap (2)
were unaffected by the
.B RLIMIT_DATA
resource limit;
since Linux 4.7, this limit is also enforced for allocations performed using
.BR mmap (2).
.PP
To avoid corruption in multithreaded applications,
mutexes are used internally to protect the memory-management
data structures employed by these functions.
In a multithreaded application in which threads simultaneously
allocate and free memory,
there could be contention for these mutexes.
To scalably handle memory allocation in multithreaded applications,
glibc creates additional
.IR "memory allocation arenas"
if mutex contention is detected.
Each arena is a large region of memory that is internally allocated
by the system
(using
.BR brk (2)
or
.BR mmap (2)),
and managed with its own mutexes.
.PP
SUSv2 requires
.BR malloc (),
.BR calloc (),
and
.BR realloc ()
to set
.I errno
to
.B ENOMEM
upon failure.
Glibc assumes that this is done
(and the glibc versions of these routines do this); if you
use a private malloc implementation that does not set
.IR errno ,
then certain library routines may fail without having
a reason in
.IR errno .
.PP
Crashes in
.BR malloc (),
.BR calloc (),
.BR realloc (),
or
.BR free ()
are almost always related to heap corruption, such as overflowing
an allocated chunk or freeing the same pointer twice.
.PP
The
.BR malloc ()
implementation is tunable via environment variables; see
.BR mallopt (3)
for details.
.SH SEE ALSO
.\" http://g.oswego.edu/dl/html/malloc.html
.\" A Memory Allocator - by Doug Lea
.\"
.\" http://www.bozemanpass.com/info/linux/malloc/Linux_Heap_Contention.html
.\" Linux Heap, Contention in free() - David Boreham
.\"
.\" http://www.citi.umich.edu/projects/linux-scalability/reports/malloc.html
.\" malloc() Performance in a Multithreaded Linux Environment -
.\"     Check Lever, David Boreham
.\"
.ad l
.nh
.BR valgrind (1),
.BR brk (2),
.BR mmap (2),
.BR alloca (3),
.BR malloc_get_state (3),
.BR malloc_info (3),
.BR malloc_trim (3),
.BR malloc_usable_size (3),
.BR mallopt (3),
.BR mcheck (3),
.BR mtrace (3),
.BR posix_memalign (3)
.SH COLOPHON
This page is part of release 4.14 of the Linux
.I man-pages
project.
A description of the project,
information about reporting bugs,
and the latest version of this page,
can be found at
\%https://www.kernel.org/doc/man\-pages/.


src/lmos-tray/parser.cpp
#include "parser.hpp"
#include <QSettings>
#include <QTime>
#include <QtCore>
/* #include "network.hpp" */
#include "pugixml.hpp"
#include "t_common.h"

#include <unistd.h>
#include <sstream>

Parser::Parser(QObject *parent) : QObject(parent)
{
	QObject::connect(&lmos, SIGNAL(log(QString)), this,
			 SLOT(log(QString)));
	QObject::connect(&lmos, SIGNAL(log(int, const char *, QString)), this,
			 SLOT(log(int, const char *, QString)));
	QObject::connect(&lmos, SIGNAL(ack(QString)), this,
			 SLOT(ack(QString)));

	QSettings traySettings("Rfbevanco", "lmos-tray");
	QString mqttsvr = traySettings.value("mqttServer").toString();
	QByteArray ba = mqttsvr.toLocal8Bit();
	const char *c_str2 = ba.data();

	blast = bp_new();
	bp_connectToServer(blast, c_str2, "lmos", 1000);
	usleep(200000);
	bp_subscribe(blast, "lmos", 1000);

	timer = new QTimer(this);
	connect(timer, SIGNAL(timeout()), this, SLOT(update()));
	timer->start(200);
}

Parser::~Parser()
{
	timer->stop();
	bp_unsubscribe(blast, "lmos", 1000);
	bp_disconnectFromServer(blast);
	bp_delete(blast);
}

void
Parser::update()
{ /* Examine the mqtt message stack */

	if (bp_getMessageCount(blast)) {
		t_bp_message msg = bp_getNewestMessage(blast);
		pugi::xml_document xml;
		xml.load_buffer(msg.data, msg.length);

		int command =
			QString(xml.child("command").child_value()).toInt();
		// qInfo() << "Parser::update() Command string: " << command;
		/* emit sendlog(command); */
		if (command) {
			parseCommand(QString(xml.child("command")
						     .attribute("id")
						     .value())
					     .toInt(),
				     command, xml);
		} else {
			// qInfo() << "Couldn't parse xml (shown below).";
			// qInfo() << QString(msg.data);
		}
	}
}

void
Parser::log(QString string)
{
	emit sendlog(string);
}

void
Parser::log(int level, const char *function, QString entry)
{
	QString time = QTime::currentTime().toString("hh:mm:ss.zzz");
	QString log = QString::number(level);
	QString logFunc = QString(function);
	emit sendlog(time + ": (" + log + ") [" + logFunc + "] " + entry);
}

void
Parser::ack(QString message)
{ /* Send a network acknowledgment */

	bp_sendMessage(blast, "broadcast", message.toStdString().c_str());
}

void
Parser::ackReturn(int id, int retval)
{ /* Send a network acknowledgment */

	QString message = QString::number(id) + "," + QString::number(retval);
	bp_sendMessage(blast, "broadcast", message.toStdString().c_str());
}

void
Parser::parseCommand(int id, int command, pugi::xml_document &xml)
{
	log("(" + QTime::currentTime().toString("hh:mm:ss.zzz") + ") " +
	    "ID: " + QString::number(id) +
	    ", Command: " + QString::number(command));

	std::stringstream out;
	QString time = QTime::currentTime().toString("hh:mm:ss.zzz");
	pugi::xml_node cmd = xml.child("command");
	QPixmap pixmap;
	QByteArray bArray;
	QBuffer buffer(&bArray);

	switch (command) {
		case kInitMachine:
			lmos.InitMachine();
			break;
		case kTermMachine:
			lmos.TermMachine();
			break;
		case kClearQpSets:
			lmos.ClearQPSets();
			lmos.SaveQPSets();
			break;
		case kImportXML:
			/* xml.child("DRAWING").remove_child("COMMAND"); */
			xml.child("command").child("DRAWING").print(
				out, "", pugi::format_raw);
			/* xml.save(out); */
			/* qInfo() << "Drawing string: " */
			/* 	<< QString::fromStdString(out.str()); */
			ackReturn(id, lmos.LoadXML(QString::fromStdString(
					      out.str())));
			break;
		case kStopPosHelp:
			lmos.StopPosHelp();
			ackReturn(id, 0);
			break;
		case kStartMarking:
			lmos.StartMarking();
			ackReturn(id, 0);
			break;
		case kStopMarking:
			lmos.StopMarking();
			ackReturn(id, 0);
			break;
		case kSelfTest:
			lmos.Test();
			break;
		case kStartPosHelp:
			lmos.StartPosHelp(cmd.attribute("object").value());
			ackReturn(id, 0);
			break;
		case kExit:
			emit finished();
			ackReturn(id, 0);
			break;
		case kReference:
			lmos.Reference();
			ackReturn(id, 0);
			break;
		case kSetMOLayer:
			lmos.StartPosHelp(cmd.attribute("object").value());
			ackReturn(id, 0);
			break;
		case kLayerSetHeight:
			lmos.LayerSetHeightZAxis(
				cmd.attribute("layer").value(),
				QString(cmd.attribute("height").value())
					.toFloat());
			ackReturn(id, 0);
			break;
		case kLayerSetLaserable:
			lmos.LayerSetLaserable(
				cmd.attribute("layer").value(),
				QString(cmd.attribute("laserable").value())
					.toInt());
			ackReturn(id, 0);
			break;
		case kLayerSetVisible:
			lmos.LayerSetVisible(
				cmd.attribute("layer").value(),
				QString(cmd.attribute("visible").value())
					.toInt());
			ackReturn(id, 0);
			break;
		case kLayerSetExportable:
			lmos.LayerSetExportable(
				cmd.attribute("layer").value(),
				QString(cmd.attribute("exportable").value())
					.toInt());
			ackReturn(id, 0);
			break;
		case kCancelJob:
			lmos.CancelJob();
			ackReturn(id, 0);
			break;
		case kSetDimension:
			lmos.SetDimension(cmd.attribute("object").value(),
					  QString(cmd.attribute("x").value())
						  .toDouble(),
					  QString(cmd.attribute("y").value())
						  .toDouble());
			ackReturn(id, 0);
			break;
		case kSetPosValues:
			lmos.SetPosValues(cmd.attribute("object").value(),
					  QString(cmd.attribute("x").value())
						  .toDouble(),
					  QString(cmd.attribute("y").value())
						  .toDouble(),
					  QString(cmd.attribute("r").value())
						  .toDouble());
			ackReturn(id, 0);
			break;
		case kZoomWindow:
			lmos.ZoomWindow(
				QString(cmd.attribute("x1").value()).toInt(),
				QString(cmd.attribute("y1").value()).toInt(),
				QString(cmd.attribute("x2").value()).toInt(),
				QString(cmd.attribute("y2").value()).toInt());
			break;
		case kShowMarkingArea:
			lmos.ShowMarkingArea();
			break;
		case kAddQpSet:
			/* qInfo() << "Parsing qpsets"; */
			for (pugi::xml_node qpset =
				     xml.child("command").child("qpset");
			     qpset; qpset = qpset.next_sibling("qpset")) {
				lmos.AddQPSet(
					qpset.attribute("name").value(),
					QString(qpset.attribute("current")
							.value())
						.toDouble(),
					QString(qpset.attribute("speed")
							.value())
						.toInt(),
					QString(qpset.attribute("frequency")
							.value())
						.toInt());
			}
			/* qInfo() << "Finished parsing qpsets"; */
			lmos.SaveQPSets();
			ackReturn(id, 0);
			break;
		case kWriteIOBit:
			lmos.WriteIOBit(
				cmd.attribute("bitfunction").value(),
				QString(cmd.attribute("value").value())
					.toInt());
			ackReturn(id, 0);
			break;
		/* case kReadByte: */
		/* 	lmos.ReadByte(getXmlInt(xml, "port"), */
		/* 		      getXmlInt(xml, "mask")); */
		/* 	break; */
		/* case kReadIOBit: */
		/* 	lmos.ReadIOBit(getXml(xml, "bitfunction")); */
		/* 	break; */
		case kSaveVLM:
			lmos.SaveVLM(cmd.attribute("filename").value());
			ackReturn(id, 0);
			emit settray(
				"Saving as " +
				QString(cmd.attribute("filename").value()));
			break;
		case kLoadVLM:
			lmos.LoadVLM(
				QString(cmd.attribute("filename").value()));
			break;
		case kMoveZ:
			lmos.MoveZ(QString(cmd.attribute("height").value())
					   .toFloat());
			ackReturn(id, 0);
			break;
		case kMoveW:
			lmos.MoveW(QString(cmd.attribute("rotation").value())
					   .toDouble());
			break;
		case kGetPng:
			pixmap = lmos.GrabWindow();
			buffer.open(QIODevice::WriteOnly);
			pixmap.save(&buffer, "PNG");
			bArray = bArray.toBase64();
			ack(QString::number(id) + "," + bArray);
			break;
		case kNoCommand: /* Dummy command for unit tests */
		case 99:	 /* Dummy command for unit tests */
			ackReturn(id, 9);
			break;

		default:
			// We should not get here
			qDebug() << "Unknown command " << command
				 << " in parseCommand()";
	}
}

void
Parser::receiveEvent(QString event)
{  // Signal has been received from Lmos

	qDebug() << "myEvent: " << event;
}
